### DISCUSSION TITLE
shell script launch

### DISCUSSION THREAD

**Discussion by anonyme-x22** [autokey/autokey#736] [General]
Hi,

I've made a script which i would like to start at launch time. (without launching autokey)
What is the comand line for this ?

Thanks in advance,
Have a good day,
Rgds,

**Comment by josephj11** [comment]
Welcome to AutoKey!

We have an autokey-run program to launch AutoKey events from external programs such as shell scripts, but it requires AutoKey to be running. It sends AutoKey a command to run using dbus. I have never used it.

AutoKey's main claim to fame is its ability to launch events via trigger abbreviations and hotkeys. If you have an application that doesn't require this capability, it would probably be cleaner and easier to program it directly in bash or in the programming language of your choice (including Python). Instead of using our API, you could use invocations of the xdotool program or the (partially) corresponding xdo Python module. There are several other similar tools, but xdotool is the only one I have experience with. I use it to configure my KDE desktop/applications when I login. You might also want to look at the tkinter tools for Python.

**Comment by anonyme-x22** [comment]
Hi,

Thanks for the answer, i've tried to use python for lauching the script, but don't figured how to import the autokey api.
i just want to start a program and move the windows.

Thanks in advance,
Rgds,

**Comment by josephj11** [comment]
Xdotool will do that. You may have to install it on your system.

Both KDE and Gnome (and probably other desktop environments) have a startup directory. Scripts (and other executables) placed there will run when the desktop starts. I don't use Gnome, but KDE tends to run these scripts early, so you may have to start your script with a delay of a second or two to give it time to finish doing whatever setup it does. Or, if there's some condition you can test for, you can use that to see if the desktop is ready for whatever your script needs to do.

Xdotool can search for existing windows, activate them and move them around the screen or to different virtual desktops. If you're opening a new window, you might have to check that it wasn't already restored from the previous session to avoid duplicating it.

Xdotool takes a bit of getting used to. It has this really cool feature called command chaining. If you start with a search command (or something else), it will put the window ID it finds in a stack. Your next command (on the same command line) will default to using that window. So you can find/open a window and then move it to the appropriate desktop all in one command.

All the details of this are beyond the scope of AutoKey, but we can pursue this on my personal Gitter page, etc. If you want to do this, then let me know how to contact you and I'll send you an invite.
